odxtools/cli/_print_utils.py

```diff
@@ -53,31 +72,208 @@ def print_diagnostic_service(
+    if not getattr(service.request, "required_parameters", None):
+        if len(service()) > 24:
+            print(
+                f"   Byte-Array: {service()[:24]!r}...\tHex-String: 0x{service().hex().upper()[:24]}..."
+            )
+        else:
+            print(f"   Byte-Array: {service()!r}\tHex-String: 0x{service().hex().upper()}")
+    else:
+        print(f"   Byte-Array: ---\tHex-String: 0x---")
+    print(f"   Service Parameters: {service.request.parameters}\n")
+    table = generatetable('parameter-details', list(service.request.parameters))
+    print(tabulate(table, headers='keys', tablefmt='presto'))
+    print("\n")
+    print(f"   Message format of the request:")
+    service.request.print_message_format(
+        indent=0, allow_unknown_lengths=allow_unknown_bit_lengths, print_objects=print_objects)
 
-            print(f"  Message format of a negative response:")
-            resp.print_message_format(indent=3, allow_unknown_lengths=allow_unknown_bit_lengths)
+    # Positive Response
+    print("\n")
+    print(f"  Positive Response Properties:")
+    print(f"   Number of Positive Responses: {len(service.positive_responses)}")
+    print(f"   Positive Responses: {service.positive_responses}")
+    if len(service.positive_responses) == 1:
+        resp = service.positive_responses[0]
+        print(f"   Service Parameters: {resp.parameters}\n")
+        table = generatetable('parameter-details', list(resp.parameters))
+        print(tabulate(table, headers='keys', tablefmt='presto'))
+        print("\n")
+        print(f"   Message format of the positive response:")
+        resp.print_message_format(
+            indent=0, allow_unknown_lengths=allow_unknown_bit_lengths, print_objects=print_objects)
+
+    # Negative Response
+    print("\n")
+    print(f"  Negative Response Properties:")
+    print(f"   Number of Negative Responses: {len(service.negative_responses)}")
+    print(f"   Negative Responses: {service.negative_responses}")
+    if len(service.negative_responses) == 1:
+        resp = service.negative_responses[0]
+        print(f"   Service Parameters: {resp.parameters}\n")
+        table = generatetable('parameter-details', list(resp.parameters))
+        print(tabulate(table, headers='keys', tablefmt='presto'))
+        print("\n")
+        print(f"   Message format of a negative response:")
+        resp.print_message_format(
+            indent=0, allow_unknown_lengths=allow_unknown_bit_lengths, print_objects=print_objects)
 
     if (service.positive_responses and
             len(service.positive_responses) > 1) or (service.negative_responses and
                                                      len(service.negative_responses) > 1):
         # Does this ever happen?
         raise NotImplementedError(
             f"The diagnostic service {service.odx_id} offers more than one response!")
+
+
+def generatetable(style: str, list_of_items: Union[List[DiagService],
```

Comment:
you are reinventing rich tables https://rich.readthedocs.io/en/stable/tables.html

Reply:
as far as I can see, this function is a bit of a misnomer: it does not produce the ASCII art itself, but only extracts the data expected by [`tabulate.tabulate()`](https://pypi.org/project/tabulate/) (which indeed seems to be a simpler version of the "rich" module). 

```suggestion
def extract_tabulation_data(style: str, list_of_items: Union[List[DiagService],
```

Besides that, what I'd like to see is to make the `style` argument an enum, so that there is less room for error, i.e.,
```suggestion
class TabulationMode(Enum):
    SERVICE_DETAILS = "service-details"
    PARAMETER_DETAILS = "parameter-details"

def extract_tabulation_data(mode: TabulationMode, list_of_items: Union[List[DiagService],
```

Even better (?), the function could be split:
```python
def extract_service_tabulation_data(services: List[DiagService]) -> Dict[str, Any]:
    ...

def extract_parameter_tabulation_data(parameters: List[Parameter]) -> Dict[str, Any]:
    ...
```

Reply:
The functionality of `generatetable` was split into `extract_service_tabulation_data` and `extract_parameter_tabulation_data` as suggested. It might be reasonable to switch from tabulate to rich tables in the future as the rich module is already used for formatting the text output in the terminal.